tests/bgp/test_bgp_peer_shutdown.py

```diff
@@ -0,0 +1,191 @@
+    else:
+        _constants.log_dir = None
+
+    return _constants
+
+
+def is_neighbor_session_established(duthost, neighbor):
+    # handle both multi-sic and single-asic
+    bgp_facts = duthost.bgp_facts(num_npus=duthost.sonichost.num_asics())["ansible_facts"]
+    return (neighbor.ip in bgp_facts["bgp_neighbors"]
+            and bgp_facts["bgp_neighbors"][neighbor.ip]["state"] == "established")
+
+
+def bgp_notification_packets(pcap_file):
+    """Get bgp notification packets from pcap file."""
+    packets = sniff(
+        offline=pcap_file,
+        lfilter=lambda p: IP in p and bgp.BGPHeader in p and p[bgp.BGPHeader].type == 3,
+    )
+    return packets
+
+
+def match_bgp_notification(packet, src_ip, dst_ip, action):
+    """Check if the bgp notification packet matches."""
+    if not (packet[IP].src == src_ip and packet[IP].dst == dst_ip):
+        return False
+
+    bgp_fields = packet[bgp.BGPNotification].fields
+    if action == "cease":
+        # error_code 6: Cease, error_subcode 2: Administrative Shutdown. References: RFC 4271
+        return bgp_fields["error_code"] == 6 and bgp_fields["error_subcode"] == 2
+    else:
+        return False
+
+
+def is_neighbor_session_down(duthost, neighbor_ip):
+    # handle both multi-sic and single-asic
+    bgp_neighbors = duthost.bgp_facts(num_npus=duthost.sonichost.num_asics())["ansible_facts"]["bgp_neighbors"]
+    return (neighbor_ip in bgp_neighbors and
+            bgp_neighbors[neighbor_ip]["admin"] == "down" and
+            bgp_neighbors[neighbor_ip]["state"] == "idle")
+
+
+def test_bgp_peer_shutdown(
+    common_setup_teardown,
+    constants,
+    duthosts,
+    enum_rand_one_per_hwsku_frontend_hostname,
+    request,
+):
+    duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
+
+    n0 = common_setup_teardown
+    try:
+        n0.start_session()
+
+        # ensure new sessions are ready
+        if not wait_until(
+            WAIT_TIMEOUT,
+            5,
+            20,
+            lambda: is_neighbor_session_established(duthost, n0),
+        ):
+            pytest.fail("Could not establish bgp sessions")
+
+        # Send BGP shutdown from DUT to n0
+        bgp_pcap = BGP_DOWN_LOG_TMPL
+        with capture_bgp_packages_to_file(duthost, "any", bgp_pcap, n0.namespace):
+            duthost.shell("config bgp shutdown neighbor {}".format(n0.name))
+            time.sleep(constants.sleep_interval)
+
+        local_pcap_filename = fetch_and_delete_pcap_file(bgp_pcap, constants.log_dir, duthost, request)
+        bpg_notifications = bgp_notification_packets(local_pcap_filename)
+        for bgp_packet in bpg_notifications:
+            logging.debug(
+                "bgp notification packet, capture time %s, packet details:\n%s",
+                bgp_packet.time,
+                bgp_packet.show(dump=True),
+            )
+
+            match_bgp_notification(bgp_packet, n0.peer_ip, n0.ip, "cease")
```

Comment:
Should we trigger the shutdown event from it's neighbor? then check when DUT receives the cease pkt and session goes to down. In correct sequence, cease pkt received, session is down, all routes are cleared.

Reply:
Good idea, triggering the shutdown event from its neighbor would allow us to test how the DUT responds to a BGP shutdown event. I will make the change.